webapp/pages/settings/my-social-media.spec.js

```diff
@@ -43,79 +38,161 @@ describe('my-social-media.vue', () => {
         wrapper = Wrapper()
-        const socialMediaLink = wrapper.find('a').attributes().href
-        expect(socialMediaLink).toBe(socialMediaUrl)
+        form = wrapper.find('form')
       })
 
-      beforeEach(() => {
-        mocks = {
-          $t: jest.fn(),
-          $apollo: {
-            mutate: jest
-              .fn()
-              .mockRejectedValue({ message: 'Ouch!' })
-              .mockResolvedValueOnce({
-                data: { DeleteSocialMeda: { id: 's1', url: socialMediaUrl } },
-              }),
-          },
-          $toast: {
-            error: jest.fn(),
-            success: jest.fn(),
-          },
-        }
-        getters = {
-          'auth/user': () => {
-            return {
-              socialMedia: [{ id: 's1', url: socialMediaUrl }],
-            }
-          },
-        }
+      describe('for each link it', () => {
+        it('displays the favicon', () => {
+          expect(wrapper.find(`img[src="${faviconUrl}"]`).exists()).toBe(true)
+        })
+
+        it('displays the url', () => {
+          expect(wrapper.find(`a[href="${socialMediaUrl}"]`).exists()).toBe(true)
+        })
+
+        it('displays the edit button', () => {
+          expect(wrapper.find('a[name="edit"]').exists()).toBe(true)
+        })
+
+        it('displays the delete button', () => {
+          expect(wrapper.find('a[name="delete"]').exists()).toBe(true)
+        })
       })
 
-      it('displays a trash sympol after a social media and allows the user to delete it', () => {
-        wrapper = Wrapper()
-        const deleteSelector = wrapper.find({ name: 'delete' })
-        expect(deleteSelector).toEqual({ selector: 'Component' })
-        const icon = wrapper.find({ name: 'trash' })
-        icon.trigger('click')
-        expect(mocks.$apollo.mutate).toHaveBeenCalledTimes(1)
+      it('does not accept a duplicate url', () => {
+        input = wrapper.find('input#addSocialMedia')
+
+        input.setValue(socialMediaUrl)
+        form.trigger('submit')
+
+        expect(mocks.$apollo.mutate).not.toHaveBeenCalled()
       })
-    })
 
-    describe('currentUser does not have a social media account linked', () => {
-      it('allows a user to add a social media link', () => {
-        wrapper = Wrapper()
-        input = wrapper.find({ name: 'social-media' })
-        input.element.value = socialMediaUrl
-        input.trigger('input')
-        submitBtn = wrapper.find('.ds-button')
-        submitBtn.trigger('click')
-        expect(mocks.$apollo.mutate).toHaveBeenCalledTimes(1)
+      describe('editing social media link', () => {
+        beforeEach(() => {
+          const editButton = wrapper.find('a[name="edit"]')
+          editButton.trigger('click')
+          input = wrapper.find('input#editSocialMedia')
+        })
+
+        it('disables adding new links while editing', () => {
+          const addInput = wrapper.find('input#addSocialMedia')
+
+          expect(addInput.exists()).toBe(false)
```

Comment:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-07-17T11:27:47Z" title="Wednesday, July 17th 2019, 1:27:47 pm +02:00">Jul 17, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -43,79 +38,152 @@ describe('my-social-media.vue', () => {
   })
 
   describe('mount', () => {
+    let form, input, submitButton
     const Wrapper = () => {
-      store = new Vuex.Store({
+      const store = new Vuex.Store({
         getters,
       })
       return mount(MySocialMedia, { store, mocks, localVue })
     }
 
-    it('renders', () => {
-      wrapper = Wrapper()
-      expect(wrapper.contains('div')).toBe(true)
+    describe('adding social media link', () => {
+      beforeEach(() => {
+        wrapper = Wrapper()
+        form = wrapper.find('form')
+        input = wrapper.find('input#addSocialMedia')
+        submitButton = wrapper.find('button')
+      })
+
+      it('requires the link to be a valid url', () => {
+        input.setValue('some value')
+        form.trigger('submit')
+
+        expect(mocks.$apollo.mutate).not.toHaveBeenCalled()
+      })
+
+      it('displays an error message when not saved successfully', async () => {
+        mocks.$apollo.mutate.mockRejectedValue({ message: 'Ouch!' })
+        input.setValue(newSocialMediaUrl)
+        form.trigger('submit')
+
+        await flushPromises()
+
+        expect(mocks.$toast.error).toHaveBeenCalledTimes(1)
+      })
+
+      describe('success', () => {
+        beforeEach(() => {
+          mocks.$apollo.mutate.mockResolvedValue({
+            data: { CreateSocialMeda: { id: 's2', url: newSocialMediaUrl } },
+          })
+          input.setValue(newSocialMediaUrl)
+          form.trigger('submit')
+        })
+
+        it('sends the new url to the backend', () => {
+          const expected = expect.objectContaining({
+            variables: { url: newSocialMediaUrl },
+          })
+
+          expect(mocks.$apollo.mutate).toHaveBeenCalledTimes(1)
+          expect(mocks.$apollo.mutate).toHaveBeenCalledWith(expected)
+        })
+
+        it('displays a success message', async () => {
+          await flushPromises()
+
+          expect(mocks.$toast.success).toHaveBeenCalledTimes(1)
+        })
+
+        it('clears the form', async () => {
+          await flushPromises()
+
+          expect(input.value).toBe(undefined)
+          expect(submitButton.vm.$attrs.disabled).toBe(true)
+        })
+      })
     })
 
-    describe('given currentUser has a social media account linked', () => {
+    describe('given existing social media links', () => {
       beforeEach(() => {
         getters = {
-          'auth/user': () => {
-            return {
-              socialMedia: [{ id: 's1', url: socialMediaUrl }],
-            }
-          },
+          'auth/user': () => ({
+            socialMedia: [{ id: 's1', url: socialMediaUrl }],
+          }),
         }
-      })
 
-      it("displays a link to the currentUser's social media", () => {
         wrapper = Wrapper()
-        const socialMediaLink = wrapper.find('a').attributes().href
-        expect(socialMediaLink).toBe(socialMediaUrl)
+        form = wrapper.find('form')
       })
 
-      beforeEach(() => {
-        mocks = {
-          $t: jest.fn(),
-          $apollo: {
-            mutate: jest
-              .fn()
-              .mockRejectedValue({ message: 'Ouch!' })
-              .mockResolvedValueOnce({
-                data: { DeleteSocialMeda: { id: 's1', url: socialMediaUrl } },
-              }),
-          },
-          $toast: {
-            error: jest.fn(),
-            success: jest.fn(),
-          },
-        }
-        getters = {
-          'auth/user': () => {
-            return {
-              socialMedia: [{ id: 's1', url: socialMediaUrl }],
-            }
-          },
-        }
+      it('displays the links', () => {
+        expect(wrapper.find(`img[src="${faviconUrl}"]`).exists()).toBe(true)
+        expect(wrapper.find(`a[href="${socialMediaUrl}"]`).exists()).toBe(true)
+        expect(wrapper.find('a[name="edit"]').exists()).toBe(true)
+        expect(wrapper.find('a[name="delete"]').exists()).toBe(true)
       })
 
-      it('displays a trash sympol after a social media and allows the user to delete it', () => {
-        wrapper = Wrapper()
-        const deleteSelector = wrapper.find({ name: 'delete' })
-        expect(deleteSelector).toEqual({ selector: 'Component' })
-        const icon = wrapper.find({ name: 'trash' })
-        icon.trigger('click')
-        expect(mocks.$apollo.mutate).toHaveBeenCalledTimes(1)
+      it('does not accept a duplicate url', () => {
+        input = wrapper.find('input#addSocialMedia')
+
+        input.setValue(socialMediaUrl)
+        form.trigger('submit')
+
+        expect(mocks.$apollo.mutate).not.toHaveBeenCalled()
       })
-    })
 
-    describe('currentUser does not have a social media account linked', () => {
-      it('allows a user to add a social media link', () => {
-        wrapper = Wrapper()
-        input = wrapper.find({ name: 'social-media' })
-        input.element.value = socialMediaUrl
-        input.trigger('input')
-        submitBtn = wrapper.find('.ds-button')
-        submitBtn.trigger('click')
-        expect(mocks.$apollo.mutate).toHaveBeenCalledTimes(1)
+      describe('editing social media link', () => {
+        beforeEach(() => {
+          const editButton = wrapper.find('a[name="edit"]')
+          editButton.trigger('click')
+          input = wrapper.find('input#editSocialMedia')
+        })
+
+        it('disables adding new links while editing', () => {
+          const addInput = wrapper.find('input#addSocialMedia')
+
+          expect(addInput.exists()).toBe(false)
```


this is great!